benchmark/Bench.hs

```diff
@@ -105,39 +111,91 @@ main_highlevel filename = do
   let size      = fileSize status
       lastBlock :: Int
       lastBlock = fromIntegral (size `div` 4096 - 1)
-      nbufs     = 64 * 4
       params    = IOCtxParams {
                     ioctxBatchSizeLimit   = 64,
                     ioctxConcurrencyLimit = 64 * 4
                   }
-      blocks    = V.fromList $ zip [0..] (randomPermute rng [0..lastBlock])
-  bracket (initIOCtx params) closeIOCtx $ \ioctx -> do
-    buf <- newPinnedByteArray (4096 * nbufs)
-
-    before <- getCurrentTime
-    forConcurrently_ (groupsOfN 32 blocks) $ \batch ->
-      submitIO ioctx $ flip fmap batch $ \ (i, block) ->
-        let bufOff  = (i `mod` nbufs) * 4096
-            blockoff = fromIntegral (block * 4096)
-        in  IOOpRead fd blockoff buf bufOff 4096
-    after <- getCurrentTime
-    let total   = lastBlock + 1
-    report before after total
-
-report :: UTCTime -> UTCTime -> Int -> IO ()
-report before after total = do
-    putStrLn $ "Total I/O ops: " ++ show total
-    putStrLn $ "Elapsed time:  " ++ show elapsed
-    putStrLn $ "IOPS:          " ++ show iops
+      ntasks    = 4
+      batchsz   = 32
+      nbatches  = lastBlock `div` (ntasks * batchsz) -- batches per task
+      totalOps  = nbatches * batchsz * ntasks
+  bracket (initIOCtx params) closeIOCtx $ \ioctx ->
+    withReport totalOps $ do
+      tasks <-
+        forRngSplitM ntasks rng $ \ !rng_task ->
+          Async.async $ do
+            buf <- newPinnedByteArray (4096 * batchsz)
+            forRngSplitM_ nbatches rng_task $ \ !rng_batch ->
+              submitIO ioctx $
+                generateIOOpsBatch fd buf lastBlock batchsz rng_batch
+      _ <- Async.waitAnyCancel tasks
+      return ()
+
+{-# NOINLINE generateIOOpsBatch #-}
+generateIOOpsBatch :: Posix.Fd
+                   -> MutableByteArray RealWorld
+                   -> Int
+                   -> Int
+                   -> Random.StdGen
+                   -> V.Vector (IOOp IO)
+generateIOOpsBatch !fd !buf !lastBlock !size !rng0 =
+    V.create $ do
+      v <- VM.new size
+      go v rng0 0
+      return v
+  where
+    go :: V.MVector s (IOOp IO) -> Random.StdGen -> Int -> ST s ()
+    go !_ !_   !i | i == size = return ()
+    go !v !rng !i = do
+      let (!block, !rng') = Random.uniformR (0, lastBlock) rng
```

Comment:
Just to clarify, previously we read each block exactly once, now it's random, so some will be read multiple times, some never? That should be fine for this benchmark.